preview/git.js

```diff
@@ -0,0 +1,252 @@
+  diffLastCommit: branch => {
+    const stream = streamChild(child_process.spawn(
+      "git",
+      ["diff-tree", "-z", "--find-renames", "--numstat", branch, "--"],
+      {
+        cwd: dir,
+        /*
+         * We use the magic 'buffer' encoding so we don't have to build a
+         * a string out of the whole thing at once. We have convenient nuls
+         * in the parsing process that we can use to "chunk" this.
+         */
+        encoding: 'buffer',
+      }
+    ));
+    return parseDiffTreeZ(stream[Symbol.asyncIterator]());
+  },
+  /**
+   * Turn a spawned child process into a stream containing its stdout and
+   * emitting an error if it fails. Exported for testing only.
+   */
+  _streamChild: streamChild,
+  /**
+   * Parse the output of `git diff-tree -z --find-renames --numstat` as an
+   * async generator. Exported for testing only.
+   */
+  _parseDiffTreeZ: parseDiffTreeZ,
+}};
+
+const streamChild = (child) => {
+  // Error should be fairly short so we can safely spool them into a variable.
+  let stderrBuffer = '';
+  child.stderr.setEncoding('utf8');
+  child.stderr.addListener('data', chunk => {
+    stderrBuffer += chunk;
+  });
+
+  let flushCallback;
+  const out = child.stdout.pipe(new Transform({
+    transform(chunk, _encoding, callback) {
+      callback(null, chunk);
+    },
+    flush(callback) {
+      // Wait to emit the end until the process closes.
+      flushCallback = callback;
+    }
+  }));
+
+  let closed = false;
+  child.addListener('close', (code, signal) => {
+    /*
+     * We can get this call multiple times for some reason. Lets just ignore
+     * the second one.....
+     */
+    if (closed) {
+      return;
+    }
+    closed = true;
+    /*
+     * Since we've closed stdout we can be sure that our transform stream has
+     * received its `flush` callback. So we delegate to that now to close
+     * the transform stream with the results of the subprocess.
+     */
+    if (code) {
+      /*
+       * Normalize some "not found" style errors from git so the caller can
+       * 404 on them.
+       */
+      let missing = stderrBuffer.includes("Not a valid object name");
+      missing |= stderrBuffer.includes("fatal: bad revision");
+      if (missing) {
+        flushCallback("missing");
+      } else {
+        flushCallback(failureMessage(`Child failed with code ${code}`, stderrBuffer));
+      }
+    } else if (signal) {
+      flushCallback(failureMessage(`Child died with signal ${signal}`, stderrBuffer));
+    } else {
+      flushCallback();
+    }
+  });
+
+  return out;
+}
+
+const failureMessage = (firstPart, stderr) => {
+  if (stderr) {
+    return `${firstPart} and stderr:\n${stderr}`;
+  }
+  return firstPart;
+}
+
+const parseDiffTreeZ = async function* (itr) {
+  const loadFirstChunk = await itr.next();
+  if (loadFirstChunk.done) {
+    // Empty diff!
+    return;
+  }
+  let chunk = loadFirstChunk.value;
+  const sliceOffNul = async from => {
+    while (true) {
```

Comment:
Did the option of using a [Duplex stream](https://nodejs.org/api/stream.html#stream_class_stream_duplex) come up instead of an infinite loop? It does add a little complexity to a simpler use case like this, but it'd be taking better advantage of a stdlib tool that's designed for handling this exact kind of need.

Reply:
I had a Transform stream implementation of this and it required a whole bunch of accounting that was difficult to get right. To the point where I'm pretty sure I didn't. The async-generators feel a little less tricky to write. Though how they execute is a bit of a trip.

Reply:
Fair point. As I was reading I wondered if it was something like that. They're awesome when you get them dialed in just right, but it takes some time to do the dialing-in well. :+1: